File: paper/src/main/java/com/github/aecsocket/sokol/paper/SokolCommand.java
package com.github.aecsocket.sokol.paper;

import cloud.commandframework.ArgumentDescription;

import cloud.commandframework.arguments.standard.EnumArgument;
import cloud.commandframework.arguments.standard.IntegerArgument;
import cloud.commandframework.arguments.standard.StringArgument;
import cloud.commandframework.bukkit.parsers.PlayerArgument;
import cloud.commandframework.bukkit.parsers.selector.MultiplePlayerSelectorArgument;
import cloud.commandframework.context.CommandContext;
import com.github.aecsocket.minecommons.core.Colls;
import com.github.aecsocket.minecommons.core.Components;
import com.github.aecsocket.minecommons.core.i18n.I18N;
import com.github.aecsocket.minecommons.core.i18n.Renderable;
import com.github.aecsocket.minecommons.core.node.NodePath;
import com.github.aecsocket.minecommons.paper.plugin.BaseCommand;
import com.github.aecsocket.sokol.core.Tree;
import com.github.aecsocket.sokol.core.context.Context;
import com.github.aecsocket.sokol.core.registry.Keyed;
import com.github.aecsocket.sokol.core.stat.StatIntermediate;
import com.github.aecsocket.sokol.core.stat.StatMap;
import com.github.aecsocket.sokol.core.world.ItemCreationException;
import com.github.aecsocket.sokol.paper.command.BlueprintArgument;
import com.github.aecsocket.sokol.paper.command.BlueprintNodeArgument;
import com.github.aecsocket.sokol.paper.command.ComponentArgument;
import com.github.aecsocket.sokol.paper.context.PaperContext;
import com.github.aecsocket.sokol.paper.world.PaperItemUser;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.JoinConfiguration;
import net.kyori.adventure.text.event.ClickEvent;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.List;
import java.util.Locale;

import static net.kyori.adventure.text.Component.*;

/* package */ final class SokolCommand extends BaseCommand<SokolPlugin> {
    private static final JoinConfiguration NEWLINE = JoinConfiguration.separator(newline());
    private static final List<String> REGISTRY_TYPES = List.of(
        PaperComponent.class.getSimpleName(),
        PaperBlueprint.class.getSimpleName()
    );

    public static final String
        KEYED_HOVER = "keyed.hover",
        FEATURE_DESCRIPTION = "feature_description",
        ERROR_ITEM_CREATION = "error.item_creation",
        ERROR_ITEM_NOT_TREE = "error.item_not_tree",
        COMMAND_LIST_ENTRY = "command.list.entry",
        COMMAND_LIST_TOTAL = "command.list.total",
        COMMAND_INFO_HEADER = "command.info.header",
        COMMAND_INFO_DESCRIPTION = "command.info.description",
        COMMAND_COMPONENT_TAGS = "command.component.tags",
        COMMAND_COMPONENT_SLOTS = "command.component.slots",
        COMMAND_COMPONENT_SLOT = "command.component.slot",
        COMMAND_COMPONENT_FEATURES = "command.component.features",
        COMMAND_COMPONENT_FEATURE = "command.component.feature",
        COMMAND_COMPONENT_STATS = "command.component.stats",
        COMMAND_COMPONENT_STAT = "command.component.stat",
        COMMAND_TREE_ROOT = "command.tree.root",
        COMMAND_TREE_INDENT = "command.tree.indent",
        COMMAND_TREE_EMPTY = "command.tree.empty",
        COMMAND_TREE_CHILD = "command.tree.child",
        COMMAND_TREE_CHILD_HOVER = "command.tree.child_hover",
        COMMAND_TREE_STATS = "command.tree.stats",
        COMMAND_TREE_INCOMPLETE_HEADER = "command.tree.incomplete.header",
        COMMAND_TREE_INCOMPLETE_ENTRY = "command.tree.incomplete.entry",
        COMMAND_GIVE = "command.give";

    public SokolCommand(SokolPlugin plugin) throws Exception {
        super(plugin, "sokol",
                (mgr, root) -> mgr.commandBuilder(root, ArgumentDescription.of("Plugin main command.")));

        registerCaption(BlueprintNodeArgument.ARGUMENT_PARSE_FAILURE_BLUEPRINT_NODE_REGISTRY);
        registerCaption(BlueprintNodeArgument.ARGUMENT_PARSE_FAILURE_BLUEPRINT_NODE_GENERIC);
        registerCaption(ComponentArgument.ARGUMENT_PARSE_FAILURE_COMPONENT_REGISTRY);
        registerCaption(BlueprintArgument.ARGUMENT_PARSE_FAILURE_BLUEPRINT_REGISTRY);

        manager.command(root
            .literal("list", ArgumentDescription.of("List all registered objects."))
            .flag(manager.flagBuilder("type").withAliases("t")
                .withArgument(StringArgument.newBuilder("value")
                    .quoted()
                    .withSuggestionsProvider((ctx, inp) -> REGISTRY_TYPES))
                .withDescription(ArgumentDescription.of("The type of object to get.")))
            .flag(manager.flagBuilder("filter").withAliases("f")
                .withArgument(StringArgument.newBuilder("value")
                    .quoted()
                    .withSuggestionsProvider((ctx, inp) -> Colls.joinList(plugin.components().keySet(), plugin.blueprints().keySet())))
                .withDescription(ArgumentDescription.of("The filter for the name or ID.")))
            .permission(permission("list"))
            .handler(c -> handle(c, this::list)));
        manager.command(root
            .literal("component", ArgumentDescription.of("Shows information on a component."))
            .argument(ComponentArgument.of(plugin, "object"), ArgumentDescription.of("The component to list information on."))
            .permission(permission("component"))
            .handler(c -> handle(c, this::component)));
        manager.command(root
            .literal("blueprint", ArgumentDescription.of("Shows information on a blueprint."))
            .argument(BlueprintArgument.of(plugin, "object"), ArgumentDescription.of("The blueprint to list information on."))
            .permission(permission("blueprint"))
            .handler(c -> handle(c, this::blueprint)));
        manager.command(root
            .literal("tree", ArgumentDescription.of("Shows information on the currently held node tree."))
            .argument(EnumArgument.optional(EquipmentSlot.class, "slot"), ArgumentDescription.of("The slot to get the item from."))
            .argument(PlayerArgument.optional("target"), ArgumentDescription.of("The player to get the item from."))
            .permission(permission("tree"))
            .handler(c -> handle(c, this::tree)));
        manager.command(root
            .literal("give", ArgumentDescription.of("Gives a blueprint tree to players."))
            .argument(MultiplePlayerSelectorArgument.of("targets"), ArgumentDescription.of("The players to give the item to."))
            .argument(BlueprintNodeArgument.of(plugin, "node"), ArgumentDescription.of("The blueprint tree to give."))
            .argument(IntegerArgument.<CommandSender>newBuilder("amount").withMin(1).asOptional(),
                ArgumentDescription.of("The amount of the item to give."))
            .permission(permission("give"))
            .handler(c -> handle(c, this::give)));
        manager.command(root
            .literal("build", ArgumentDescription.of("Builds a registered blueprint and gives it to players."))
            .argument(MultiplePlayerSelectorArgument.of("targets"), ArgumentDescription.of("The players to give the item to."))
            .argument(BlueprintArgument.of(plugin, "blueprint"), ArgumentDescription.of("The blueprint to build."))
            .argument(IntegerArgument.<CommandSender>newBuilder("amount").withMin(1).asOptional(),
                ArgumentDescription.of("The amount of the item to give."))
            .permission(permission("build"))
            .handler(c -> handle(c, this::build)));
    }

    private <T extends Renderable & Keyed> Component renderKeyedHover(Locale locale, T obj, String command) {
        return join(NEWLINE, i18n.lines(locale, KEYED_HOVER,
            c -> c.of("id", () -> text(obj.id())),
            c -> c.of("type", () -> text(obj.getClass().getName())),
            c -> c.of("command", () -> text(command))));
    }

    private String command(PaperComponent object) {
        return "/%s component %s".formatted(rootName, object.id());
    }

    private String command(PaperBlueprint object) {
        return "/%s blueprint %s".formatted(rootName, object.id());
    }

    private @Nullable String command(Keyed object) {
        return object instanceof PaperComponent cast
            ? command(cast)
            : object instanceof PaperBlueprint cast
            ? command(cast)
            : null;
    }

    private void list(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        String type = ctx.flags().get("type");
        if (type != null) type = type.toLowerCase(Locale.ROOT);
        String filter = ctx.flags().get("filter");
        if (filter != null) filter = filter.toLowerCase(Locale.ROOT);

        List<Keyed> registered = Colls.joinList(plugin.components().values(), plugin.blueprints().values());
        int results = 0;
        for (var object : registered) {
            if (type != null) {
                String typeName = object.getClass().getName().toLowerCase(Locale.ROOT);
                if (!typeName.contains(type))
                    continue;
            }
            Component rendered = object.render(i18n, locale);
            String id = object.id();
            if (filter != null) {
                if (
                    !id.toLowerCase(Locale.ROOT).contains(filter)
                    && !PlainTextComponentSerializer.plainText().serialize(rendered).toLowerCase(Locale.ROOT).contains(filter)
                )
                    continue;
            }

            ++results;
            String command = command(object);
            Component hover = command == null ? null : renderKeyedHover(locale, object, command);
            ClickEvent click = command == null ? null : ClickEvent.runCommand(command);
            plugin.send(sender, i18n.modLines(locale, COMMAND_LIST_ENTRY,
                line -> line.hoverEvent(hover).clickEvent(click),
                c -> c.of("type", () -> text(object.getClass().getSimpleName())),
                c -> c.of("name", () -> c.rd(object)),
                c -> c.of("id", () -> text(id))));
        }

        int fResults = results;
        plugin.send(sender, i18n.lines(locale, COMMAND_LIST_TOTAL,
            c -> c.of("amount", () -> text(fResults))));
    }

    private void component(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        PaperComponent object = ctx.get("object");

        info(ctx, sender, locale, pSender, object);

        plugin.send(sender, i18n.lines(locale, COMMAND_COMPONENT_TAGS,
            c -> c.of("tags", () -> text(String.join(", ", object.tags())))));

        plugin.send(sender, i18n.lines(locale, COMMAND_COMPONENT_SLOTS,
            c -> c.of("amount", () -> text(object.slots().size()))));
        for (var entry : object.slots().entrySet()) {
            String id = entry.getKey();
            PaperNodeSlot slot = entry.getValue();

            Component hover = slot.rule().render(i18n, locale);
            plugin.send(sender, i18n.modLines(locale, COMMAND_COMPONENT_SLOT,
                line -> line.hoverEvent(hover),
                c -> c.of("name", () -> c.rd(slot)),
                c -> c.of("id", () -> text(id)),
                c -> c.of("tags", () -> text(String.join(", ", slot.tags()))),
                c -> c.of("offset", () -> text(""+slot.offset()))));
        }

        plugin.send(sender, i18n.lines(locale, COMMAND_COMPONENT_FEATURES,
            c -> c.of("amount", () -> text(object.features().size()))));
        for (var entry : object.features().entrySet()) {
            String id = entry.getKey();
            PaperFeatureProfile<?, ?> profile = entry.getValue();

            Component hover = profile.type().renderDescription(i18n, locale)
                .map(lines -> Component.join(NEWLINE, lines.stream()
                    .map(line -> i18n.line(locale, FEATURE_DESCRIPTION,
                        c -> c.of("line", () -> line)))
                    .toList()))
                .orElse(null);
            plugin.send(sender, i18n.modLines(locale, COMMAND_COMPONENT_FEATURE,
                line -> line.hoverEvent(hover),
                c -> c.of("name", () -> c.rd(profile.type())),
                c -> c.of("id", () -> text(id))));
        }

        List<StatIntermediate.MapData> allData = object.stats().join();
        plugin.send(sender, i18n.lines(locale, COMMAND_COMPONENT_STATS,
            c -> c.of("amount", () -> text(allData.size()))));
        for (var data : allData) {
            StatMap stats = data.entries();

            Component hover = join(NEWLINE, stats.render(i18n, locale));
            plugin.send(sender, i18n.modLines(locale, COMMAND_COMPONENT_STAT,
                line -> line.hoverEvent(hover),
                c -> c.of("amount", () -> text(stats.size())),
                c -> c.of("priority", () -> c.rd(data.priority())),
                c -> c.of("rule", () -> c.rd(data.rule()))));
        }
    }

    private void blueprint(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        PaperBlueprint object = ctx.get("object");

        info(ctx, sender, locale, pSender, object);

        tree(ctx, sender, locale, pSender, object.create(), 1);
    }

    private void info(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender, Keyed object) {
        plugin.send(sender, i18n.lines(locale, COMMAND_INFO_HEADER,
            c -> c.of("type", () -> text(object.getClass().getSimpleName())),
            c -> c.of("name", () -> c.rd(object)),
            c -> c.of("id", () -> text(object.id()))));

        object.renderDescription(i18n, locale).ifPresent(desc -> {
            for (var line : desc) {
                plugin.send(sender, i18n.lines(locale, COMMAND_INFO_DESCRIPTION,
                    c -> c.of("line", () -> line)));
            }
        });
    }

    private void tree(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        EquipmentSlot slot = ctx.getOrDefault("slot", EquipmentSlot.HAND);
        Player target = defaultedArg(ctx, "target", pSender, p -> p);

        @SuppressWarnings("ConstantConditions")
        ItemStack item = target.getInventory().getItem(slot);
        PaperBlueprintNode node = plugin.persistence().load(item)
            .orElseThrow(() -> error(ERROR_ITEM_NOT_TREE));

        tree(ctx, sender, locale, pSender, node, 0);
    }

    private void subTree(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender, Component indent, int depth, PaperBlueprintNode node) {
        for (var entry : node.value().slots().entrySet()) {
            String key = entry.getKey();
            PaperNodeSlot slot = entry.getValue();
            node.get(slot.key()).ifPresentOrElse(child -> {
                String command = command(child.value());
                I18N.TemplateFactory[] templates = new I18N.TemplateFactory[] {
                    c -> c.of("indent", () -> Components.repeat(indent, depth)),
                    c -> c.of("slot", () -> c.rd(slot)),
                    c -> c.of("key", () -> text(key)),
                    c -> c.of("name", () -> c.rd(child.value())),
                    c -> c.of("id", () -> text(child.value().id()))
                };
                Component hover = i18n.line(locale, COMMAND_TREE_CHILD_HOVER, templates)
                    .append(renderKeyedHover(locale, child.value(), command));
                ClickEvent click = ClickEvent.runCommand(command);
                plugin.send(sender, i18n.modLines(locale, COMMAND_TREE_CHILD,
                    line -> line.hoverEvent(hover).clickEvent(click),
                    templates));
                subTree(ctx, sender, locale, pSender, indent, depth + 1, child);
            }, () -> plugin.send(sender, i18n.lines(locale, COMMAND_TREE_EMPTY,
                c -> c.of("indent", () -> Components.repeat(indent, depth)),
                c -> c.of("slot", () -> c.rd(slot)),
                c -> c.of("key", () -> text(key)))));
        }
    }

    private void tree(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender, PaperBlueprintNode node, int depth) {
        String command = command(node.value());
        Component indent = i18n.line(locale, COMMAND_TREE_INDENT);
        Component rootHover = renderKeyedHover(locale, node.value(), command);
        ClickEvent rootClick = ClickEvent.runCommand(command);
        plugin.send(sender, i18n.modLines(locale, COMMAND_TREE_ROOT,
            line -> line.hoverEvent(rootHover).clickEvent(rootClick),
            c -> c.of("indent", () -> Components.repeat(indent, depth)),
            c -> c.of("name", () -> c.rd(node.value())),
            c -> c.of("id", () -> text(node.value().id()))));
        subTree(ctx, sender, locale, pSender, indent, depth, node);

        Tree<PaperTreeNode> tree = node.asTreeNode(Context.context(locale)).tree();
        StatMap stats = tree.stats();
        List<NodePath> incomplete = tree.incomplete();

        Component statsHover = Component.join(NEWLINE, stats.render(i18n, locale));
        plugin.send(sender, i18n.modLines(locale, COMMAND_TREE_STATS,
            line -> line.hoverEvent(statsHover),
            c -> c.of("amount", () -> text(stats.size()))));
        plugin.send(sender, i18n.lines(locale, COMMAND_TREE_INCOMPLETE_HEADER,
            c -> c.of("amount", () -> text(incomplete.size()))));
        for (var path : incomplete) {
            plugin.send(sender, i18n.lines(locale, COMMAND_TREE_INCOMPLETE_ENTRY,
                c -> c.of("path", () -> text(String.join("/", path)))));
        }
    }

    private void give(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender, PaperBlueprintNode blueprint) {
        ItemStack baseItem;
        try {
            baseItem = blueprint.asTreeNode(
                pSender == null
                    ? Context.context(locale)
                    : PaperContext.context(PaperItemUser.user(plugin, pSender))
            ).asItem().handle();
        } catch (ItemCreationException e) {
            throw error(ERROR_ITEM_CREATION, e);
        }

        @SuppressWarnings("ConstantConditions")
        int amount = ctx.getOrDefault("amount", 1);
        List<Player> targets = targets(ctx, "targets", pSender);

        for (var target : targets) {
            ItemStack item = blueprint.asTreeNode(PaperContext.context(PaperItemUser.user(plugin, target)))
                .asItem().handle();
            Inventory inventory = target.getInventory();
            for (int i = 0; i < amount; i++) {
                inventory.addItem(item);
            }
        }

        plugin.send(sender, i18n.lines(locale, COMMAND_GIVE,
            c -> c.of("amount", () -> text(amount)),
            c -> c.of("item", baseItem::displayName),
            c -> c.of("targets", () -> targets.size() == 1
                ? targets.get(0).displayName()
                : text(targets.size()))));
    }

    private void give(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        give(ctx, sender, locale, pSender, ctx.get("node"));
    }

    private void build(CommandContext<CommandSender> ctx, CommandSender sender, Locale locale, @Nullable Player pSender) {
        give(ctx, sender, locale, pSender, ctx.<PaperBlueprint>get("blueprint").create());
    }
}

File: paper/src/main/kotlin/com/github/aecsocket/sokol/paper/SokolCommand.kt
package com.github.aecsocket.sokol.paper

import cloud.commandframework.arguments.standard.BooleanArgument
import cloud.commandframework.arguments.standard.EnumArgument
import cloud.commandframework.arguments.standard.IntegerArgument
import cloud.commandframework.arguments.standard.StringArgument
import cloud.commandframework.bukkit.arguments.selector.MultiplePlayerSelector
import cloud.commandframework.bukkit.parsers.selector.MultiplePlayerSelectorArgument
import cloud.commandframework.context.CommandContext
import com.github.aecsocket.alexandria.core.extension.*
import com.github.aecsocket.alexandria.core.keyed.Keyed
import com.github.aecsocket.alexandria.core.keyed.Registry
import com.github.aecsocket.alexandria.paper.plugin.CloudCommand
import com.github.aecsocket.alexandria.paper.plugin.desc
import com.github.aecsocket.glossa.core.Localizable
import com.github.aecsocket.sokol.core.BlueprintParser
import com.github.aecsocket.sokol.core.ComponentParser
import com.github.aecsocket.sokol.core.NodeParser
import net.kyori.adventure.extra.kotlin.join
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.Component.text
import net.kyori.adventure.text.JoinConfiguration
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import org.bukkit.command.CommandSender
import org.bukkit.craftbukkit.v1_18_R2.entity.CraftAreaEffectCloud
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.event.entity.CreatureSpawnEvent
import org.spigotmc.AsyncCatcher.enabled
import java.util.*

internal class SokolCommand(plugin: SokolPlugin) : CloudCommand<SokolPlugin>(
    plugin, "sokol",
    { manager, rootName -> manager.commandBuilder(rootName, desc("Core plugin command.")) }
) {
    init {
        captions?.apply {
            registerMessageFactory(ComponentParser.ARGUMENT_PARSE_FAILURE_COMPONENT, captionLocalizer)
            registerMessageFactory(BlueprintParser.ARGUMENT_PARSE_FAILURE_BLUEPRINT, captionLocalizer)
            registerMessageFactory(NodeParser.ARGUMENT_PARSE_FAILURE_DATA_NODE_REGISTRY, captionLocalizer)
            registerMessageFactory(NodeParser.ARGUMENT_PARSE_FAILURE_DATA_NODE_MALFORMED, captionLocalizer)
        }

        val hosts = root
            .literal("hosts", desc("Gets info on the last hosts resolved on the server."))
        manager.command(hosts
            .permission(perm("hosts"))
            .handler { handle(it, ::hosts) })
        manager.command(hosts
            .literal("toggle", desc("Toggle the hosts HUD element."))
            .permission(perm("hosts", "toggle"))
            .senderType(Player::class.java)
            .handler { handle(it, ::hostsToggle) })

        val list = root
            .literal("list", desc("Lists various registered items."))
        manager.command(list
            .literal("features", desc("Lists all registered features."))
            .argument(StringArgument.optional("filter"), desc("Filter for the name or ID, case-insensitive."))
            .permission(perm("list", "features"))
            .handler { handle(it, ::listFeatures) })
        manager.command(list
            .literal("components", desc("Lists all registered components."))
            .argument(StringArgument.optional("filter"), desc("Filter for the name or ID, case-insensitive."))
            .permission(perm("list", "components"))
            .handler { handle(it, ::listComponents) })
        manager.command(list
            .literal("blueprints", desc("Lists all registered blueprints."))
            .argument(StringArgument.optional("filter"), desc("Filter for the name or ID, case-insensitive."))
            .permission(perm("list", "blueprints"))
            .handler { handle(it, ::listBlueprints) })

        val info = root
            .literal("info", desc("Gets detailed info on a specific registered item."))
        manager.command(info
            .literal("component", desc("Gets info on a registered component."))
            .argument(PaperComponentArgument(plugin, "item", desc("Item to get info for.")))
            .permission(perm("info", "component"))
            .handler { handle(it, ::infoComponent) })
        manager.command(info
            .literal("blueprint", desc("Gets info on a registered blueprint."))
            .argument(PaperBlueprintArgument(plugin, "item", desc("Item to get info for.")))
            .permission(perm("info", "blueprint"))
            .handler { handle(it, ::infoBlueprint) })

        val render = root
            .literal("render", desc("Options for rendered items,"))
        manager.command(render
            .literal("show-shapes", desc("Toggles showing shape and point definitions on the selected render."))
            .argument(EnumArgument.optional(NodeRender.ShowShape::class.java, "state"), desc("For what objects to show shapes of."))
            .permission(perm("render", "show-shapes"))
            .senderType(Player::class.java)
            .handler { handle(it, ::renderShowShapes) })
        /*manager.command(render
            .literal("rotate", desc("Clears the currently set  view rotation."))
            .permission(perm("inspect", "rotate"))
            .senderType(Player::class.java)
            .handler { handle(it, ::renderRotate) })
        manager.command(render
            .literal("rotate", desc("Rotates all inspect views by the specified angle."))
            .argumentEuler3("rotation", desc("Angle to rotate to, in Euler degrees."))
            .argument(EnumArgument.optional(EulerOrder::class.java, "order"), desc("Order for the Euler angles."))
            .permission(perm("inspect", "rotate"))
            .senderType(Player::class.java)
            .handler { handle(it, ::renderRotateSet) }) todo */

        manager.command(root
            .literal("give", desc("Gives a specified item-representable node tree to a player."))
            .argument(MultiplePlayerSelectorArgument.of("targets"), desc("Players to give to."))
            .argument(PaperNodeArgument(plugin, "item", desc("Node to give.")))
            .argument(IntegerArgument.optional("amount"), desc("Amount of the item to give."))
            .permission(perm("give"))
            .handler { handle(it, ::give) })
        manager.command(root
            .literal("build", desc("Builds and gives a specified item-representable blueprint to a player."))
            .argument(MultiplePlayerSelectorArgument.of("targets"), desc("Players to give to."))
            .argument(PaperBlueprintArgument(plugin, "item", desc("Blueprint to give.")))
            .argument(IntegerArgument.optional("amount"), desc("Amount of the item to give."))
            .permission(perm("build"))
            .handler { handle(it, ::build) })
    }

    fun <T : Keyed> byRegistry(registry: Registry<T>, id: String, locale: Locale) = registry[id]
        ?: error { safe(locale, "error.invalid_id") {
            raw("id") { id }
        } }

    fun component(id: String, locale: Locale) = byRegistry(plugin.components, id, locale)

    fun blueprint(id: String, locale: Locale) = byRegistry(plugin.blueprints, id, locale)


    fun hosts(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        val hosts = plugin.lastHosts
        val possible = hosts.values.sumOf { it.possible }
        val marked = hosts.values.sumOf { it.marked }
        plugin.send(sender) { safe(locale, "command.hosts") {
            raw("possible") { possible }
            raw("marked") { marked }
            raw("percent") { marked.toDouble() / possible }
            list("types") { hosts.map { (type, data) ->
                map {
                    raw("name") { type }
                    raw("possible") { data.possible }
                    raw("marked") { data.marked }
                    raw("percent") { data.marked.toDouble() / data.possible }
                }
            } }
        } }
    }

    fun hostsToggle(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        plugin.playerData(sender as Player).apply { showHosts = !showHosts }
    }

    fun <T> list(
        ctx: CommandContext<CommandSender>,
        sender: CommandSender,
        locale: Locale,
        registry: Registry<T>
    ) where T : Keyed, T : Localizable<Component> {
        val filter = ctx.getOrDefault<String>("filter", null)

        fun List<Component>.contains(string: String) = any {
            PlainTextComponentSerializer.plainText().serialize(it).contains(string)
        }

        val results = ArrayList<T>()
        registry.forEach { (_, item) ->
            if (filter?.let {
                item.id.contains(filter)
                    || item.localize(plugin.i18n.withLocale(locale)).contains(filter)
            } != false) {
                results.add(item)
            }
        }

        plugin.send(sender) { safe(locale, "command.list") {
            raw("found") { results.size }
            raw("total") { registry.size }
            list("entries") { results.forEach { item ->
                map {
                    tl("name") { item }
                    raw("id") { item.id }
                }
            } }
        } }
    }

    fun listFeatures(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        list(ctx, sender, locale, plugin.features)
    }

    fun listComponents(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        list(ctx, sender, locale, plugin.components)
    }

    fun listBlueprints(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        list(ctx, sender, locale, plugin.blueprints)
    }


    fun infoComponent(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        val component = ctx.get<PaperComponent>("item")
        plugin.send(sender) { safe(locale, "command.info.component.message") {
            tl("name") { component }
            raw("id") { component.id }
            list("tags") { component.tags.forEach {
                raw(it)
            } }
            raw("qt_slots") { component.slots.size }
            list("slots") { component.slots.forEach { (key, slot) ->
                val hover = text("") // todo

                subList(safe("command.info.component.slot") {
                    tl("name") { slot }
                    raw("key") { key }
                    raw("required") { slot.required.toString() }
                    raw("modifiable") { slot.modifiable.toString() }
                    list("tags") { slot.tags.forEach {
                        raw(it)
                    } }
                }.map { it.hoverEvent(hover) })
            } }
            raw("qt_features") { component.features.size }
            list("features") { component.features.forEach { (key, feature) ->
                val hover = component.featureConfigs[key]!!.render()
                    .join(JoinConfiguration.newlines())

                subList(safe("command.info.component.feature") {
                    tl("name") { feature.type }
                    raw("id") { key }
                }.map { it.hoverEvent(hover) })
            } }
            raw("qt_stats") { component.stats.size }
            list("stats") { component.stats.forEach { stats ->
                // todo hover

                subList(safe("command.info.component.stats") {
                    raw("priority") { stats.priority }
                    raw("reversed") { stats.reversed.toString() }
                    raw("qt_entries") { stats.stats.entries.size }
                })
            } }
        } }
    }

    fun infoTree(sender: CommandSender, locale: Locale, node: PaperDataNode): List<Component> {
        val res = ArrayList<Component>()

        res.addAll(plugin.i18n.safe(locale, "command.tree.header") {
            tl("name") { node.component }
            raw("id") { node.component.id }
        })

        fun visitSlot(slot: PaperSlot, node: PaperDataNode?, depth: Int) {
            res.addAll(plugin.i18n.safe(locale, "command.tree.child") {
                raw("indent") { "  ".repeat(depth) }
                tl("slot") { slot }
                subList("value") {
                    node?.let {
                        node.component.localize(this)
                    } ?: safe("command.tree.empty_slot")
                }
            })

            node?.let {
                node.component.slots.forEach { (key, child) ->
                    visitSlot(child, node.node(key), depth + 1)
                }
            }
        }

        node.component.slots.forEach { (key, slot) ->
            visitSlot(slot, node.node(key), 0)
        }

        return res
    }

    fun infoBlueprint(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        val blueprint = ctx.get<PaperBlueprint>("item")
        plugin.send(sender) { safe(locale, "command.info.blueprint") {
            tl("name") { blueprint }
            raw("id") { blueprint.id }
            subList("tree") { infoTree(sender, locale, blueprint.createNode()) }
        } }
    }

    fun renderShowShapes(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        val data = plugin.playerData(sender as Player)
        val state = ctx.get("state") { NodeRender.ShowShape.NONE }

        data.rdShowShapes = state
        plugin.send(sender) { safe(locale, "command.render.show_shapes.${state.key}") }
    }

    fun renderRotate(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        val data = plugin.playerData(sender as Player)
        // todo data.isRotation = null
    }

    fun renderRotateSet(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        val data = plugin.playerData(sender as Player)
        val rotation = ctx.get<Euler3>("rotation")
        val order = ctx.get("order") { EulerOrder.XYZ }

        val quaternion = rotation.radians.quaternion(order)
        // todo data.isRotation = quaternion
    }

    fun give(
        ctx: CommandContext<CommandSender>,
        sender: CommandSender,
        locale: Locale,
        node: PaperDataNode
    ) {
        val targets = ctx.get<MultiplePlayerSelector?>("targets") { null }.orSender("targets", sender, locale)
        val amount = ctx.get("amount") { 1 }

        val state = paperStateOf(node)
        val stack = try {
            plugin.persistence.stateToStack(state)
        } catch (ex: NodeItemCreationException) {
            error(ex) { safe(locale, "error.creating_item") }
        }

        targets.forEach { target ->
            (0 until amount).forEach { _ -> target.inventory.addItem(stack) }
        }

        plugin.send(sender) {
            if (targets.size == 1) {
                safe(locale, "command.give.one") {
                    sub("item") { stack.displayName() }
                    sub("target") { targets.first().displayName() }
                    raw("amount") { amount }
                }
            } else {
                safe(locale, "command.give.other") {
                    sub("item") { stack.displayName() }
                    raw("qt_targets") { targets.size }
                    raw("amount") { amount }
                }
            }
        }
    }

    fun give(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        give(ctx, sender, locale, ctx.get<PaperDataNode>("item"))
    }

    fun build(ctx: CommandContext<CommandSender>, sender: CommandSender, locale: Locale) {
        give(ctx, sender, locale, ctx.get<PaperBlueprint>("item").createNode())
    }
}
